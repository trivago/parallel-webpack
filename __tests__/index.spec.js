jest.mock('testConfig.js', () => ({}), { virtual: true });

import { run } from '../index';
import workerFarm  from 'worker-farm';
import Bluebird, {
    promisify,
    error,
    then,
    asCallback,
} from 'bluebird';

describe('index.js', () => {
    describe('run', () => {
        beforeEach(() => {
            jest.clearAllMocks();
        });
        afterEach(() => {
            process.removeAllListeners('SIGINT');
        })

        it('should reject promise config cannot get loaded', () => {
            const returnPromise = run('./does/not/exist.js', { colors: false }, jest.fn());
            expect(returnPromise.reject.mock.calls[0][0].toString()).toMatchSnapshot();
        });

        it('should reject promise if options validation fails', () => {
            const returnPromise = run('testConfig.js', { maxConcurrentWorkers: 'fail' }, jest.fn());
            expect(returnPromise.reject.mock.calls[0]).toMatchSnapshot();
        });

        it('should call generate workers and return farm promise', () => {
            jest.spyOn(console, 'log').mockImplementation(() => {});
            promisify.mockReturnValueOnce(jest.fn());

            const returnPromise = run('testConfig.js', { colors: false }, jest.fn());
            expect(workerFarm.mock.calls[0][0]).toEqual({ maxRetries: 0 });

            expect(returnPromise).toBe(Bluebird);
            expect(promisify).toHaveBeenCalledTimes(1);
            expect(error).toHaveBeenCalledTimes(1);
            expect(then).toHaveBeenCalledTimes(2);
            expect(Bluebird.finally).toHaveBeenCalledTimes(1);
            expect(asCallback).toHaveBeenCalledTimes(1);
        });

        describe('error callback', () => {
            beforeEach(() => {
                jest.spyOn(console, 'log').mockImplementation(() => {});
                jest.spyOn(Date, 'now')
                    .mockImplementationOnce(() => 0)
                    .mockImplementationOnce(() => 300);
            });

            const errorCbTest = options => {
                promisify.mockReturnValueOnce(jest.fn());

                const returnPromise = run('testConfig.js', {
                    json: options.silent,
                    colors: false
                }, jest.fn());
                const cb = returnPromise.error.mock.calls[0][0];
                const response = cb('Exception on worker farm');

                expect(response).toBe(Bluebird);
                expect(Bluebird.reject).toHaveBeenCalledWith('Exception on worker farm');
                if (options.silent) {
                    expect(console.log).not.toHaveBeenCalled();
                } else {
                    expect(console.log.mock.calls).toMatchSnapshot();
                }
            };

            it('should log and reject with error', () => {
                errorCbTest({ silent: false });
            });
            it('should only reject with error when silent', () => {
                errorCbTest({ silent: true });
            });
        });

        describe('then callback', () => {
            beforeEach(() => {
                jest.spyOn(console, 'log').mockImplementation(() => {});
                jest.spyOn(Date, 'now')
                    .mockImplementationOnce(() => 30000)
                    .mockImplementationOnce(() => 0);
            });

            const thenCbTest = options => {
                promisify.mockReturnValueOnce(jest.fn());

                const returnPromise = run('testConfig.js', {
                    json: options.silent,
                    colors: false
                }, jest.fn());
                const cb = returnPromise.then.mock.calls[1][0];
                const response = cb([true, true, false, undefined, '', 0]);

                expect(response).toEqual([ true, true ]);
                if(options.silent) {
                    expect(console.log).not.toHaveBeenCalled();
                } else {
                    expect(console.log.mock.calls).toMatchSnapshot();
                }
            };

            it('should filter non-truthy and return results', () => {
                thenCbTest({ silent: false });
            });
            it('should not log when silent', () => {
                thenCbTest({ silent: true });
            });
        });

        describe('finally callback', () => {
            it('should call end workerFarm and remove SIGINT listener', () => {
                promisify.mockReturnValueOnce(jest.fn());

                const returnPromise = run('testConfig.js', { colors: false }, jest.fn());
                const cb = returnPromise.finally.mock.calls[0][0];

                expect(process.listenerCount('SIGINT')).toBe(1);
                cb();
                expect(process.listenerCount('SIGINT')).toBe(0);

                // called with workers
                expect(workerFarm.end.mock.calls[0][0]).toBe(workerFarm.end);
            });
        });

        describe('shutdownCallback', () => {
            const shutdownTest = options => {
                promisify.mockReturnValueOnce(jest.fn());

                const returnPromise = run('testConfig.js', {
                    json: options.silent,
                    colors: false
                }, jest.fn());

                expect(process.listenerCount('SIGINT')).toBe(1);

                process.emit('SIGINT');
                // called with workers
                expect(workerFarm.end.mock.calls[0][0]).toBe(workerFarm.end);
                if(options.silent) {
                    expect(console.log).not.toHaveBeenCalled();
                } else {
                    expect(console.log.mock.calls).toMatchSnapshot();
                }
            };

            it('should call end  and remove callback', () => {
                shutdownTest({ silent: false });
            });

            it('should call end  and remove callback silently', () => {
                shutdownTest({ silent: true });

            });
        });
    });

});
