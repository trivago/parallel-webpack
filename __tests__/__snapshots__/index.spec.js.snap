// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.js run error callback should log and reject with error 1`] = `
Array [
  Array [
    "[34m[WEBPACK][39m Building [33m1[39m target",
  ],
  Array [
    "%s Build failed after %s seconds",
    "[31m[WEBPACK][39m",
    "[34m0.3[39m",
  ],
]
`;

exports[`index.js run should reject promise config cannot get loaded 1`] = `
"Error: [31m[WEBPACK][39m Could not load configuration file [4m./does/not/exist.js[24m
Error: Cannot find module './does/not/exist.js' from 'loadConfigurationFile.js'"
`;

exports[`index.js run should reject promise if options validation fails 1`] = `
Array [
  [Error: Options validation failed:
Property: "options.maxConcurrentWorkers" should be number],
]
`;

exports[`index.js run shutdownCallback should call end  and remove callback 1`] = `
Array [
  Array [
    "[34m[WEBPACK][39m Building [33m1[39m target",
  ],
  Array [
    "[31m[WEBPACK][39m Forcefully shutting down",
  ],
]
`;

exports[`index.js run then callback should filter non-truthy and return results 1`] = `
Array [
  Array [
    "[34m[WEBPACK][39m Building [33m1[39m target",
  ],
  Array [
    "%s Finished build after %s seconds",
    "[34m[WEBPACK][39m",
    "[34m29.7[39m",
  ],
]
`;
